### PR TITLE
New router/app behaviour 

### LINKED ISSUE #468: Refactor routing
Acceptance criteria:

- [x] Replace Memory router by custom router that will update `?help` query param on every route change, so you can always forward up-to-date link to specific article

How to test:

- as you navigate through the app, `?help` param in url should be in sync with your current position in the application, back / forward browser buttons should still work & after a full page refresh, the exact same page in the page should be loaded

### PR DESCRIPTION
**Pull request by dinodsaurus**
It is quite simple in the end but it took me some time to get there 🤯
So there are two main points in this new router:
* if its a new instance of our `smart-faq` redirect if needed - `Redirector.js`
* on each route change sync our `inMemory` location with `help` url query param `urlSync.js`

Also, I made some changes in frontend repo regarding rendering and persisting our in-app state.. So now on each open/close of `smart-faq` a new instance is not being spawned, it just shows/hides the initial `smart-faq` instance.

The only problem that I'm now facing is some server-side rendering stuff on the frontend(don't think this is related to this PR, but on the server, there is no `window` so we get an error). Tomorrow we can take a look at that if you have some spare time @jaroslav-kubicek 

## Breaking change
On the frontend app, if you wanna use a link to a specific `smart-faq` article you need to add`redirect=true` query param.
**example**
`https://www.kiwi.com/en/?help=%2Ffaq%2FRkFRQ2F0ZWdvcnk6ODQ%3D%2Farticle%2FRkFRQXJ0aWNsZToxMTQ%3D&redirect=true`

closes #468 

<br/><br/><br/><url>LiveURL: https://smartfaq-url-to-queryparams.surge.sh</url>

### CODE REVIEW

**Comment by jaroslav-kubicek** [comment]
> The only problem that Im now facing is some server side rendering stuff on the frontend(dont think this is related to this PR, but on the server there is no window so we get an error). Tomorrow we can take a look at that if you have some spare time

This might be an actual issue - on a server, there is no window, so you can't use window in any function that is called during SSR. Since we have reference to the window in several places, if there are executed now after your changes, then it poses a problem and the only solution is to update the code to check for the existence of window...

File: src/App.js [hunk 1]
```diff
@@ -45,7 +45,7 @@ type Props = {|
   locale: {},
   user: User,
   loginToken: ?string,
-  initialRoute: string,
+  initialRoute: ?string, // eslint-disable-line
```
[the remark this answers is not in the document]
Comment (dinodsaurus) [reply src/App.js]: The thing is that this actually triggers the rerendering of the app (prop changed) but its never used..

**Comment by dinodsaurus** [comment]
Oh yea, the problem is that earlier the `smart-faq` didn't render when frontend started, it only got rendered when there was a `?help=` param. Thanks!

File: src/Redirector.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+class Redirector extends React.Component<Props, State> {
+  state = {
+    firstOpen: true
+  }
+  static getDerivedStateFromProps(props, state) {
+    const { firstOpen } = state;
+    const paramsString = window.location.search;
+    const params = new URLSearchParams(paramsString);
```
[the remark this answers is not in the document]
Comment (dinodsaurus) [reply src/Redirector.js]: Thanks!

**Comment by jalbertsr** [comment]
I think we can close this MR since it have been solved in https://github.com/kiwicom/smart-faq/pull/549

**Pull request closed without merging**